pkg/repository/participation.go

```diff
@@ -0,0 +1,75 @@
+package repository
+
+import (
+	"context"
+	"errors"
+	"github.com/CollActionteam/collaction_backend/internal/constants"
+	m "github.com/CollActionteam/collaction_backend/internal/models"
+	"github.com/CollActionteam/collaction_backend/models"
+	"github.com/CollActionteam/collaction_backend/pkg/repository/aws"
+	"github.com/CollActionteam/collaction_backend/utils"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
+)
+
+type Participation interface {
+	Get(ctx context.Context, userID string, crowdactionID string) (*m.ParticipationRecord, error)
+	Register(ctx context.Context, userID string, name string, crowdaction *models.Crowdaction, payload m.JoinPayload) error
+	Cancel(ctx context.Context, userID string, crowdaction *models.Crowdaction) error
+}
+
+type participation struct {
+	dbClient *aws.Dynamo
+}
+
+func NewParticipation(dynamo *aws.Dynamo) Participation {
+	return &participation{
+		dbClient: dynamo,
+	}
+}
+func (s *participation) Get(ctx context.Context, userID string, crowdactionID string) (*m.ParticipationRecord, error) {
+	pk := utils.PrefixParticipationPK_UserID + userID
+	sk := utils.PrefixParticipationSK_CrowdactionID + crowdactionID
+	item, err := s.dbClient.GetDBItem(constants.TableName, pk, sk)
+	if item == nil || err != nil {
+		return nil, err
+	}
+	var r m.ParticipationRecord
+	err = dynamodbattribute.UnmarshalMap(item, &r)
+	return &r, err
+}
+
+func (s *participation) Register(ctx context.Context, userID string, name string, crowdaction *models.Crowdaction, payload m.JoinPayload) error {
+	/* TODO Password not required when joining for MVP
+	if crowdaction.PasswordJoin != "" && crowdaction.PasswordJoin != payload.Password {
+		return fmt.Errorf("invalid password")
+	}
+	*/
+	err := models.ValidateCommitments(payload.Commitments, crowdaction.CommitmentOptions)
```

Comment:
what if move this validation into the controller?

Reply:
I don't know.
If another function calls the repository directly, this validation is necessary?
If don't, we can move without problems.

Reply:
repository exists for Logic, every side must handle its own concerns and I think validation is the gateway concern  (in this case handlers)

Reply:
done